model: Improve reporting upon moving/splitting topic.

srdeotarse opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->
This PR improves reporting upon moving a message by editing it's topic.

<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->
Partial fix for #1172 

<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [x] Manually
- [x] Existing tests (adapted, if necessary)
- [x] New tests added (for any new behavior)
- [x] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->

**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->

**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- eg.
- Waiting on #<PR>
- Blocks #<PR>
-->

**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->

https://user-images.githubusercontent.com/53873549/159926734-3f04ffb8-c35f-4836-8d76-ae7401e6fca6.mp4

neiljp commented on zulipterminal/model.py:
```diff
@@ -567,8 +567,14 @@ def update_stream_message(
             if old_topic != new_topic:
-                self.controller.report_success("You changed a message's topic.")
+                self.controller.report_success(
+                    f"You changed {no_of_messages} message's topic from #{stream_name} > {old_topic} to #{stream_name} > {new_topic}."
```
I'm not sure if we have a standard way to do this using `report_success`, but this may be clearer if we can provide a highlight style to group the stream+topic in each case.

This may require changes to existing code, so that could be a follow-up.

srdeotarse replied:
After fixing the format for reporting message upon moving topics, we can look into highlighting as well.

neiljp commented on zulipterminal/model.py:
```diff
@@ -567,8 +567,14 @@ def update_stream_message(
+            stream_name = self.index["messages"][message_id].get(
+                "display_recipient", None
+            )
+            no_of_messages = len(self.index["edited_messages"])
```
This is all the edited messages we know about, so this may work the first time, with no known edited messages, but I expect likely will grow each time and so report the wrong result.

Try testing this manually, and then you can add a test case for this and check it fails, before fixing.

srdeotarse replied:
I got your point. Therefore, I have used different method -
1. Getting all stream messages
2. Filtering messages based on topic where message is edited by user.
3. Compare message_id of edited message and topic messages to get recent_edited_messages.

neiljp commented on zulipterminal/model.py:
```diff
@@ -567,8 +567,14 @@ def update_stream_message(
         if response["result"] == "success":
             old_topic = self.index["messages"][message_id].get("subject", None)
             new_topic = request["topic"]
+            stream_name = self.index["messages"][message_id].get(
+                "display_recipient", None
+            )
```
Let's simplify using a well-named variable for the common code here.

srdeotarse replied:
get_message = self.index["messages'] for common code.

neiljp reviewed (commented):
@srdeotarse Thanks for looking at this, including giving the tests a go :+1:

This doesn't appear to pass tests (CI), which is not what you have in the PR description.

Since the re-indexing occurs outside the current method where the message occurs, it's possible we may need to do that elsewhere.

srdeotarse commented:
Rebased with current main branch and updated mypy. Still showing error below @neiljp -
```
Successfully installed autoflake-1.4 autopep8-1.6.0 flake8-4.0.1 isort-5.10.1 jedi-0.18.1 lxml-stubs-0.4.0 mypy-0.942 parso-0.8.3 pudb-2022.1.1 pycodestyle-2.8.0 pyflakes-2.4.0 pytz-2022.1 snakeviz-2.1.1 types-docutils-0.18.0 types-pygments-2.9.17 types-setuptools-57.4.11 zulip-term-0.6.0+git
(zt_venv) srdeotarse@LAPTOP-VC2MC73M:~/zulip-terminal$ make check
================================================================================
Import order (isort)
--------------------------------------------------------------------------------
All files have imports sorted according to rules.
================================================================================
Type consistency (mypy)
--------------------------------------------------------------------------------
Running mypy for `zulipterminal`.
zulipterminal/model.py:584: error: Invalid index type "Optional[int]" for "Dict[int, Set[int]]"; expected type "int"
Found 1 error in 1 file (checked 37 source files)
Running mypy for `tests`.
Success: no issues found in 15 source files
================================================================================
LINTING FAILED
make: *** [makefile:22: lint] Error 1
```

srdeotarse commented:
Rebased with current main branch and updated mypy. Still showing error below @neiljp -

```
Successfully installed autoflake-1.4 autopep8-1.6.0 flake8-4.0.1 isort-5.10.1 jedi-0.18.1 lxml-stubs-0.4.0 mypy-0.942 parso-0.8.3 pudb-2022.1.1 pycodestyle-2.8.0 pyflakes-2.4.0 pytz-2022.1 snakeviz-2.1.1 types-docutils-0.18.0 types-pygments-2.9.17 types-setuptools-57.4.11 zulip-term-0.6.0+git
(zt_venv) srdeotarse@LAPTOP-VC2MC73M:~/zulip-terminal$ make check
================================================================================
Import order (isort)
--------------------------------------------------------------------------------
All files have imports sorted according to rules.
================================================================================
Type consistency (mypy)
--------------------------------------------------------------------------------
Running mypy for `zulipterminal`.
zulipterminal/model.py:584: error: Invalid index type "Optional[int]" for "Dict[int, Set[int]]"; expected type "int"
Found 1 error in 1 file (checked 37 source files)
Running mypy for `tests`.
Success: no issues found in 15 source files
================================================================================
LINTING FAILED
make: *** [makefile:22: lint] Error 1
```

neiljp commented:
@srdeotarse Was there a reason you closed/reopened this? Generally there's no need to do so.

neiljp commented on zulipterminal/model.py:
```diff
@@ -576,10 +576,28 @@ def update_stream_message(
-                self.controller.report_success("You changed a message's topic.")
+                if recent_moved_msgs == "one":
+                    self.controller.report_success(
+                        f"You changed {recent_moved_msgs} message's topic from #{stream_name} > {old_topic} to #{stream_name} > {new_topic}."
```
This is identical except for the apostrophe placement? If so, it'll be simpler to integrate the difference into the variable, and avoid the need for a conditional here.

neiljp replied:
This was what I referred to by simplifying the code in the next review.

neiljp commented on tests/model/test_model.py:
```diff
@@ -888,24 +907,54 @@ def test_update_private_message(
+        model.index["messages"][req["message_id"]]["display_recipient"] = stream_name
+        req["propagate_mode"] = propagate_mode
+
+        recent_moved_msgs = ""
+        if propagate_mode == "change_all":
+            recent_moved_msgs = "all"
+        if propagate_mode == "change_later":
+            recent_moved_msgs = "some"
+        if propagate_mode == "change_one":
+            recent_moved_msgs = "one"
```
It's better to put the explicit text into the test cases above, than have conditional code here - particularly if it is exactly the same as the code in the application.

neiljp commented on tests/model/test_model.py:
```diff
@@ -888,24 +907,54 @@ def test_update_private_message(
         report_success = model.controller.report_success
         if result and footer_updated:
-            report_success.assert_called_once_with("You changed a message's topic.")
+            if recent_moved_msgs == "one":
```
By just using a string we can assert on an 'expected string' provided in each parametrize 'case', avoiding the need for extra conditionals here.

neiljp commented on tests/model/test_model.py:
```diff
@@ -888,24 +907,54 @@ def test_update_private_message(
+                "change_all",
                 True,
             ),
         ],
```
Given the longer test here now, it would be useful to also add inline test ids.

neiljp replied:
See my followup comment in the next review.

srdeotarse commented:
@neiljp I was trying to rebase with main to solve mypy error for Optional[int] and accidentally deleted my local branch.

neiljp reviewed (commented):
@srdeotarse Thanks for the update :+1:

Other than the comments inline, I think we could simplify the code for readability, though it is functional as it is :)

One aspect this brings up is the short time given until the message disappears in the footer. For a simple solution, we could tune this manually with an optional parameter to pass through, which might be warranted as a preparatory commit before the one you have.

I've thought previously we could try to handle this automatically based on word length and reading speed, with some minimum, but that may need more tuning than a simple prep commit.

neiljp reviewed (commented):
@srdeotarse I've mainly focused on the tests here, since once they're well-defined then refactoring the code to give the same behavior should be enforced - and good test-writing technique is a useful skill.

The first commit is good to have separate, but as per my inline comment I'd suggest you keep it very simple to simply allow passing a custom duration in the main commit, with a default of the current 3 as now. The version with a formula is worthy of it's own PR since the implementation is worthy of a separate discussion and this can be merged in the meantime.

neiljp commented on zulipterminal/core.py:
```diff
@@ -447,7 +447,9 @@ def report_success(self, text: str) -> None:
-        self.view.set_footer_text(text, "task:success", 3)
+        self.view.set_footer_text(
+            text, "task:success", float(len(text.split()) * 6 / 13)
+        )  # Average reading speed = 130 wpm
```
We're discussing this in the stream now :+1:

neiljp commented:
@srdeotarse Thanks for the contribution 👍 I made a few minor textual adjustments which you might be able to identify manually, or find using `git diff` or `git range-diff`, I but merged otherwise as-is :tada:
